Docsp 28002 arithmetic operations (#344)
mongodb/docs-java: source/fundamentals/aggregation/aggregation-expression-operations.txt

   from the Query API expression provided in each example. However,
   both expressions will produce the same aggregation result.

Arithmetic Operations
~~~~~~~~~~~~~~~~~~~~~

You can use the methods described in this section to perform arithmetic

> operations on values of type ``MqlInteger`` or ``MqlNumber``.

Reviewer: Maybe an introductory sentence that doesn't reference types would be useful here, something like:

"You can use arithmetic operations to manipulate numbers. You can perform an arithmetic operation on a value of type ``MqlInteger`` or ``MqlNumber`` using the methods described in this section."

> The following table lists the arithemetic methods
> available in the driver and corresponding expressions in the
> Query API. The methods link to API documentation and the Query API
> expressions link to Server manual documentation.

Reviewer: Suggestion: describe what users will find in the Server docs

```suggestion
The following table describes arithmetic operation methods available in the driver and corresponding expression operators in the Query API. The method names link to API documentation and the aggregation pipeline operator names link to descriptions and examples in the Server manual documentation.
```

Author: Per our discussion, this should move into the string pull request.

     - :manual:`$abs </reference/operator/aggregation/abs/>`

   * - `add() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/mql/MqlNumber.html#add(com.mongodb.client.model.mql.MqlNumber)>`__
     - :manual:`$add </reference/operator/aggregation/add/>`

   * - `divide() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/mql/MqlNumber.html#divide(com.mongodb.client.model.mql.MqlNumber)>`__
     - :manual:`$divide </reference/operator/aggregation/divide/>`

   * - `multiply() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/mql/MqlNumber.html#multiply(com.mongodb.client.model.mql.MqlNumber)>`__
     - :manual:`$multiply </reference/operator/aggregation/multiply/>`

   * - `round() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/mql/MqlNumber.html#round()>`__
     - :manual:`$round </reference/operator/aggregation/round/>`

   * - `subtract() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/mql/MqlNumber.html#subtract(com.mongodb.client.model.mql.MqlNumber)>`__
     - :manual:`$subtract </reference/operator/aggregation/subtract/>`

> Suppose you have a sales database and want to calculate the
> total cost of an individual sale.

Reviewer: ```suggestion
Suppose you need to calculate the final value of a sale from the item's cost, quantity sold, and applicable discount.
```

The first step is to multiply the ``item_cost`` and the ``quantity`` fields using
the ``multiply`` method. Then use the ``subtract`` method to apply their

> ``discount``.

Reviewer: Suggestion: Rephrase to avoid "first, then" structure. Maybe, use directive;

E.g.
To calculate the cost of items sold, multiply the ... and ... fields by using the ... method. To apply the discount, subtract the ... field from the total cost by using the ... method.

> 

Reviewer: I: code example needs intro

```suggestion
The following code shows the pipeline stage for this aggregation:
```

.. code-block:: java
   :copyable: true

   var pipeline = Arrays.asList(project(fields(
            computed("total_sale",
                     current()
                           .getInteger("item_cost")
                           .multiply(current().getInteger("quantity"))
                           .subtract(current().getInteger("discount")))
            )));

The following code provides an equivalent expression in the Query API:

.. code-block:: javascript
   :copyable: true

>    { $project: { total_sale: { $subtract: [ { $multiply: [ "$item_cost", "$quantity" ] }, "$discount" ] } } }

Reviewer: Suggestion: reformat to avoid horizontal scroll

     - :manual:`$add </reference/operator/aggregation/add/>`

   * - `divide() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/mql/MqlNumber.html#divide(com.mongodb.client.model.mql.MqlNumber)>`__
     - :manual:`$divide </reference/operator/aggregation/divide/>`

   * - `multiply() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/mql/MqlNumber.html#multiply(com.mongodb.client.model.mql.MqlNumber)>`__
     - :manual:`$multiply </reference/operator/aggregation/multiply/>`

   * - `round() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/mql/MqlNumber.html#round()>`__
     - :manual:`$round </reference/operator/aggregation/round/>`

   * - `subtract() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/mql/MqlNumber.html#subtract(com.mongodb.client.model.mql.MqlNumber)>`__
     - :manual:`$subtract </reference/operator/aggregation/subtract/>`

Suppose you need to calculate the final value of a sale and have the item's
cost, the quantity sold, and an applicable discount.

To calculate the final value, multiple the ``item_cost`` and the ``quantity``
fields using the ``multiply`` method. To apply the ``discount``, use the

> ``subtract`` method.

Reviewer: S: explain that the resulting number is projected as a new field

The following code shows the pipeline stage for this aggregation:

.. code-block:: java
   :copyable: true

>    var pipeline = Arrays.asList(project(fields(
>             computed("total_sale",
>                      current()
>                            .getInteger("item_cost")
>                            .multiply(current().getInteger("quantity"))
>                            .subtract(current().getInteger("discount")))
>             )));

Reviewer: C: reformat

```suggestion
            Arrays.asList(project(fields(
                    computed("finalCost",
                            current()
                                    .getInteger("itemPrice")
                                    .multiply(current()
                                            .getInteger("quantity"))
                                    .subtract(current()
                                            .getInteger("discount")))
            )));
```

> The following code provides an equivalent expression in the Query API:

Reviewer: from string PR

```suggestion
The following code provides an equivalent aggregation pipeline in the Query API:
```

.. code-block:: javascript
   :copyable: true

>    { $project: { total_sale: {
>          $subtract: [ {
>             $multiply: [ "$item_cost", "$quantity" ] },
>             "$discount" ] } } }

Reviewer: suggested reformat (also now includes [] to indicate it is a pipeline)

```suggestion
[ { $project: {
      finalCost: {
        $subtract: [
          { $multiply: [ "$itemPrice",
                         "$quantity" ] },
          "$discount" ] } } } ]
```

   from the Query API expression provided in each example. However,
   both expressions will produce the same aggregation result.

Arithmetic Operations
~~~~~~~~~~~~~~~~~~~~~

You can use arithmetic operations to manipulate numbers. You can perform an
arithmetic operation on a value of ``MqlInteger`` or ``MqlNumber`` using the
methods described in this section.

> The following table lists the arithemetic methods

Reviewer: ```suggestion
The following table lists the arithmetic methods
```

   * - `add() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/mql/MqlNumber.html#add(com.mongodb.client.model.mql.MqlNumber)>`__
     - :manual:`$add </reference/operator/aggregation/add/>`

   * - `divide() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/mql/MqlNumber.html#divide(com.mongodb.client.model.mql.MqlNumber)>`__
     - :manual:`$divide </reference/operator/aggregation/divide/>`

   * - `multiply() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/mql/MqlNumber.html#multiply(com.mongodb.client.model.mql.MqlNumber)>`__
     - :manual:`$multiply </reference/operator/aggregation/multiply/>`

   * - `round() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/mql/MqlNumber.html#round()>`__
     - :manual:`$round </reference/operator/aggregation/round/>`

   * - `subtract() <{+api+}/apidocs/mongodb-driver-core/com/mongodb/client/model/mql/MqlNumber.html#subtract(com.mongodb.client.model.mql.MqlNumber)>`__
     - :manual:`$subtract </reference/operator/aggregation/subtract/>`

Suppose you need to calculate the total cost of a sale from the item's
cost, the quantity sold, and an applicable discount.

> To calculate the cost of the sale, multiply the ``itemPrice`` and the ``quantity``
> fields by using the ``multiply()`` method. To apply the ``discount``, use the
> ``subtract()`` method.
> 
> The following code shows the pipeline stage for this aggregation and assigns
> cost of the sale to a new field ``totalSale``:

Reviewer: S: since the projection is the stage, it would make more sense to explain the projection in the first of these two paragraphs, something like:

```suggestion
To calculate the cost of the sale, multiply the ``itemPrice`` and the ``quantity``
fields by using the ``multiply()`` method. To apply the ``discount``, use the
``subtract()`` method. To output the result as the ``totalSale`` field, use the ``project()`` method.

The following code shows the pipeline stage for this aggregation:
```
